lib/gitlab/markdown/relative_link_filter.rb

```diff
@@ -0,0 +1,117 @@
+require 'uri'
+
+module Gitlab
+  module Markdown
+    # HTML filter that "fixes" relative links to files in a repository.
+    #
+    # Context options:
+    #   :commit
+    #   :project
+    #   :project_wiki
+    #   :requested_path
+    #   :ref
+    class RelativeLinkFilter < HTML::Pipeline::Filter
+
+      def call
+        if !project_wiki && repository.try(:exists?) && !repository.empty?
+          doc.search('a').each do |el|
+            process_link_attr el.attribute('href')
+          end
+
+          doc.search('img').each do |el|
+            process_link_attr el.attribute('src')
+          end
+        end
+
+        doc
+      end
+
+      protected
+
+      def process_link_attr(html_attr)
+        return if html_attr.blank?
+
+        uri = URI(html_attr.value)
+        if uri.relative? && uri.path.present?
+          html_attr.value = rebuild_relative_uri(uri).to_s
+        end
+      end
+
+      def rebuild_relative_uri(uri)
+        file_path = relative_file_path(uri.path)
+
+        uri.path = [
+          relative_url_root,
+          project.path_with_namespace,
+          path_type(file_path),
+          ref || 'master',  # assume that if no ref exists we can point to master
+          file_path
+        ].compact.join('/').squeeze('/').chomp('/')
```

Comment:
After `compact` removes nil values, is it even possible there would be duplicate slashes requiring `squeeze`, or a trailing slash requiring `chomp`?